### ISSUE TITLE
multiple problems when declaring inline anchor ids in headers

### ISSUE THREAD

**Issue by hossman** [asciidoctor/asciidoctor#1875]
According to the user manual (section 30.1: [Defining an Anchor](http://asciidoctor.org/docs/user-manual/#anchordef) ) ...

> In addition to being able to define anchors on sections and blocks, anchors can be defined inline where ever you can type normal text (anchors are a macro substitution). The anchors in the text get replaced with invisible anchor points in the output.
> ...
> Another example with headers (the anchor needs to be in the text):
> Defining an inline anchor ID in an header
> 
> `=== [[current]]Version 4.9`

But there are multiple problems/oddities in defining anchors in headers like this.

Example:

```
= X Title
:toc:

X X X X X X X X

== [[anchor_a0]]Header A

X X X X X X X X
```

html5 backend output (of the `<body>`)...

```
<body class="article">
<div id="header">
<h1>X Title</h1>
<div id="toc" class="toc">
<div id="toctitle">Table of Contents</div>
<ul class="sectlevel1">
<li><a href="#__a_id_anchor_a0_a_header_a"><a id="anchor_a0"></a>Header A</a></li>
</ul>
</div>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
<div class="sect1">
<h2 id="__a_id_anchor_a0_a_header_a"><a id="anchor_a0"></a>Header A</h2>
<div class="sectionbody">
<div class="paragraph">
<p>X X X X X X X X</p>
</div>
</div>
</div>
</div>
<div id="footer">
<div id="footer-text">
Last updated 2016-09-13 14:14:02 -07:00
</div>
</div>
</body>
```

Problems to note:
1. An `<a id="anchor_a0"></a>` declaration exists inside both the `<h2>` for the section, as well as in the `<li>` for that section in the `toc`
2. The `<a id="anchor_a0"></a>` inside the `<h2>` is naively used as if it were part of the header text when generate the id for `<h2>`: `<h2 id="__a_id_anchor_a0_a_header_a">`

It seems to me that in examples like this, `anchor_a0` should be used "as is" to define the `id` for the `<h2>` (ie: No auto-generated section id needed), in the same way asciidoctor would behave if it were written like so...

```
[[anchor_a0]]
== Header A
```

...either that, or section 30.1 of the user guide should be changed to stop encouraging (and should probably explicitly discourage) declaring anchors inside of headers. 

---

A few final notes:
- This issue is a superset of one of the problematic cases identified in asciidoctor/asciidoctor/issues/1874
- In the examples above I noted the HTML output created by the html5 backend because it's easy to review the final documents source/structure, but I _believe_ the same problems also occur when using the pdf backend (which is my ultimate goal) .. but i didn't test extensively.
- I've been testing via asciidoctor-ant-1.5.3.jar since that's the latest version i can find on maven.org and ultimately i need to use asciidoctor-ant

**Comment by guw** [comment]
```
[[anchor_a0]]
== Header A
```

Is this supported everywhere? It doesn't seem to work in AsciiDoctor.JS and on GitHub. Only the shorthand syntax seems to work there:

```
[#anchor_a0]
== Header A
```

**Comment by hossman** [comment]
> Is this supported everywhere? ...
It is the officially documented syntax for defining an anchor -- both block level or inline. 

If you're seeing some backends where it doesn't work, then it sounds like those should be tracked as independent issues -- it seems orthogonal to the issue reported here.

**Comment by mojavelinux** [comment]
I believe these issues are now resolved as part of 1.5.7.

> An `<a id="anchor_a0"></a>` declaration exists inside both the `<h2>` for the section, as well as in the `<li>` for that section in the toc

This was resolved by the fix for #840.

> The `<a id="anchor_a0"></a>` inside the `<h2>` is naively used as if it were part of the header text when generate the id for `<h2>`: `<h2 id="__a_id_anchor_a0_a_header_a">`

This was resolved by the fix for #794.

> It seems to me that in examples like this, anchor_a0 should be used "as is" to define the id for the `<h2>` (ie: No auto-generated section id needed)

The ID for a section can be specified inline as long as it comes at the *end* of the section title (not the beginning) (and an ID is not specified on the block attribute line).

```
=== Version 4.9 [[current]]
```

However, we're considering dropping this behavior in the future. The preferred way of defining an ID for a section is to use a block attribute line.

> section 30.1 of the user guide should be changed to stop encouraging (and should probably explicitly discourage) declaring anchors inside of headers.

I 100% agree. cc: @graphitefriction
